runtime_config_linux.go

```diff
@@ -133,57 +133,57 @@ type ThrottleDevice struct {
+	ThrottleWriteBpsDevice []*ThrottleDevice `json:"blkioThrottleWriteBpsDevice,omitempty"`
 	// IO read rate limit per cgroup per device, IO per second
-	ThrottleReadIOPSDevice []*ThrottleDevice `json:"blkioThrottleReadIOPSDevice"`
+	ThrottleReadIOPSDevice []*ThrottleDevice `json:"blkioThrottleReadIOPSDevice,omitempty"`
 	// IO write rate limit per cgroup per device, IO per second
-	ThrottleWriteIOPSDevice []*ThrottleDevice `json:"blkioThrottleWriteIOPSDevice"`
+	ThrottleWriteIOPSDevice []*ThrottleDevice `json:"blkioThrottleWriteIOPSDevice,omitempty"`
 }
 
 // Memory for Linux cgroup 'memory' resource management
 type Memory struct {
-	// Memory limit (in bytes)
-	Limit uint64 `json:"limit"`
-	// Memory reservation or soft_limit (in bytes)
-	Reservation uint64 `json:"reservation"`
-	// Total memory usage (memory + swap); set `-1' to disable swap
-	Swap uint64 `json:"swap"`
-	// Kernel memory limit (in bytes)
-	Kernel uint64 `json:"kernel"`
-	// How aggressive the kernel will swap memory pages. Range from 0 to 100. Set -1 to use system default
-	Swappiness uint64 `json:"swappiness"`
+	// Memory limit (in bytes).
+	Limit *int64 `json:"limit,omitempty"`
+	// Memory reservation or soft_limit (in bytes).
+	Reservation *int64 `json:"reservation,omitempty"`
+	// Total memory limit (memory + swap).
+	Swap *int64 `json:"swap,omitempty"`
+	// Kernel memory limit (in bytes).
+	Kernel *int64 `json:"kernel,omitempty"`
+	// How aggressive the kernel will swap memory pages. Range from 0 to 100.
+	Swappiness *int64 `json:"swappiness,omitempty"`
 }
 
 // CPU for Linux cgroup 'cpu' resource management
 type CPU struct {
-	// CPU shares (relative weight vs. other cgroups with cpu shares)
-	Shares uint64 `json:"shares"`
-	// CPU hardcap limit (in usecs). Allowed cpu time in a given period
-	Quota uint64 `json:"quota"`
-	// CPU period to be used for hardcapping (in usecs). 0 to use system default
-	Period uint64 `json:"period"`
-	// How many time CPU will use in realtime scheduling (in usecs)
-	RealtimeRuntime uint64 `json:"realtimeRuntime"`
-	// CPU period to be used for realtime scheduling (in usecs)
-	RealtimePeriod uint64 `json:"realtimePeriod"`
-	// CPU to use within the cpuset
-	Cpus string `json:"cpus"`
-	// MEM to use within the cpuset
-	Mems string `json:"mems"`
+	// CPU shares (relative weight (ratio) vs. other cgroups with cpu shares).
+	Shares *int64 `json:"shares,omitempty"`
+	// CPU hardcap limit (in usecs). Allowed cpu time in a given period.
+	Quota *int64 `json:"quota,omitempty"`
+	// CPU period to be used for hardcapping (in usecs).
+	Period *int64 `json:"period,omitempty"`
+	// How much time realtime scheduling may use (in usecs).
+	RealtimeRuntime *int64 `json:"realtimeRuntime,omitempty"`
+	// CPU period to be used for realtime scheduling (in usecs).
+	RealtimePeriod *int64 `json:"realtimePeriod,omitempty"`
+	// CPUs to use within the cpuset. Default is to use any CPU available.
+	Cpus string `json:"cpus,omitempty"`
```

Comment:
Its software. Everything is easy :) But does that mean we should do it?

On Wed, Dec 16, 2015 at 11:10 AM, W. Trevor King <notifications@github.com>
wrote:

> In runtime_config_linux.go
> <https://github.com/opencontainers/specs/pull/233#discussion_r47818436>:
>
> > -	// CPU to use within the cpuset
> > -	Cpus string `json:"cpus"`
> > -	// MEM to use within the cpuset
> > -	Mems string `json:"mems"`
> > +	// CPU shares (relative weight (ratio) vs. other cgroups with cpu shares).
> > +	Shares *int64 `json:"shares,omitempty"`
> > +	// CPU hardcap limit (in usecs). Allowed cpu time in a given period.
> > +	Quota *int64 `json:"quota,omitempty"`
> > +	// CPU period to be used for hardcapping (in usecs).
> > +	Period *int64 `json:"period,omitempty"`
> > +	// How much time realtime scheduling may use (in usecs).
> > +	RealtimeRuntime *int64 `json:"realtimeRuntime,omitempty"`
> > +	// CPU period to be used for realtime scheduling (in usecs).
> > +	RealtimePeriod *int64 `json:"realtimePeriod,omitempty"`
> > +	// CPUs to use within the cpuset. Default is to use any CPU available.
> > +	Cpus string `json:"cpus,omitempty"`
>
> On Wed, Dec 16, 2015 at 09:59:24AM -0800, Vish Kannan wrote: > + // CPUs
> to use within the cpuset. Default is to use any CPU available. > + Cpus
> string `json:"cpus,omitempty"` With integer types we don't have the option
> to differentiate between `null` and `0`. But with string types, that is not
> an issue right?
> That assumes that the empty string will never be semantically meaningful
> (i.e. you'll never want to actually write the empty string to a cgroups
> pseudo-file). That may be a valid assumption, but I see no point in relying
> on it when a string pointer here would be easy to add.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/specs/pull/233/files#r47818436>.
>